Python/ceval_gil.c

```diff
@@ -206,58 +206,56 @@ static void recreate_gil(struct _gil_runtime_state *gil)
 #endif
 
 static void
-drop_gil_impl(struct _gil_runtime_state *gil)
-{
-    MUTEX_LOCK(gil->mutex);
-    _Py_ANNOTATE_RWLOCK_RELEASED(&gil->locked, /*is_write=*/1);
-    _Py_atomic_store_int_relaxed(&gil->locked, 0);
-    COND_SIGNAL(gil->cond);
-    MUTEX_UNLOCK(gil->mutex);
-}
-
-static void
-drop_gil(PyInterpreterState *interp, PyThreadState *tstate)
+drop_gil(PyInterpreterState *interp, PyThreadState *tstate, int thread_dying)
```

Comment:
Unfortunately `thread_dying` is not equivalent to `tstate == NULL`. There are three cases:
1. `tstate != NULL`, `thread_dying == 0`: normal case.
2. `tstate != NULL`, `thread_dying == 1`: thread dying normally, from [this call](https://github.com/python/cpython/blob/82abe75e77129bebb3c13d807e8040f6924194f6/Python/pystate.c#L1837). This is the "new" case, since the thread is going to die but we still need `tstate` to know if the thread holds the GIL.
3. `tstate == NULL`, `thread_dying == 1`: `_PyThreadState_MustExit` case, from [this call](https://github.com/python/cpython/blob/82abe75e77129bebb3c13d807e8040f6924194f6/Python/ceval_gil.c#L398), and what your comment is about.

I can change it to `final_drop`, since that should accurately capture the last two scenarios, even once the thread is changed to sleep forever.

Reply:
Thanks for explaining!